micov/cli.py

```diff
@@ -392,36 +392,106 @@ def _load_db(dbbase, sample_metadata, features_to_keep):
+    help="Genome lengths")
+@click.option("--outdir", type=click.Path(exists=False), required=True,
+    help="Output directory for results",)
+@click.option("--bin-num", type=int, default=1000,
+    help="Number of bins (default: 1000)")
+@click.option("--rank", is_flag=True, default=False,
+    help="Enable ranking (default: False)")
+def binning(
+    covered_positions,
+    sample_metadata,
+    features_to_keep,
+    metadata_variable,
+    length,
+    outdir,
+    bin_num,
+    rank
+):
     """Bin genome positions and quantify read and sample hits across bins."""
-    pos = pl.read_csv(covered_positions, separator='\t',
-                      new_columns=BED_COV_SAMPLEID_SCHEMA.columns,
-                      schema_overrides=BED_COV_SAMPLEID_SCHEMA.dtypes_dict,
-                      has_header=False, skip_rows=1).lazy()
-    pos = pos.filter(pl.col(COLUMN_GENOME_ID).is_in([genome_id]))
+    df_md = parse_sample_metadata(sample_metadata).lazy()
+    df_length = parse_genome_lengths(length).lazy()
+    df_pos = pl.scan_parquet(covered_positions)
+
+    df_pos_md = df_pos.join(
+        df_md, on=COLUMN_SAMPLE_ID, how="left"
+    ).join(
+        df_length, on=COLUMN_GENOME_ID, how="left"
+    )
 
-    bin_df, pos_updated = pos_to_bins(pos, genome_length, bin_num)
+    if features_to_keep:
+        features_to_keep = _first_col_as_set(features_to_keep)
+        df_pos_md = df_pos_md.filter(
+            pl.col(COLUMN_GENOME_ID).is_in(features_to_keep))
+
+    df_bins_list = []
+    genome_ids = df_pos_md.select(
+        COLUMN_GENOME_ID).unique().collect().to_series().to_list()
+    for genome_id in genome_ids:
+        pos = df_pos_md.filter(pl.col(COLUMN_GENOME_ID) == genome_id)
+        df_bins = pos_to_bins(pos, metadata_variable, bin_num)
+        df_bins_list.append(df_bins)
+
+    df_bins = pl.concat(df_bins_list)
+
+    df_bins_by_sample_hits = (
+        df_bins.group_by("genome_id", "bin_idx", "bin_start", "bin_stop")
+        .agg(pl.col("sample_hits").std().alias("sample_hits_std"))
+        .fill_null(0)
+        .sort("sample_hits_std", descending=True)
+    )
 
-    if not os.path.exists(outdir):
-        os.makedirs(outdir)
+    df_bins_by_read_hits = (
+        df_bins.group_by("genome_id", "bin_idx", "bin_start", "bin_stop")
+        .agg(pl.col("read_hits").std().alias("read_hits_std"))
+        .fill_null(0)
+        .sort("read_hits_std", descending=True)
+    )
 
-    bin_df = make_csv_ready(bin_df)
-    pos_updated = make_csv_ready(pos_updated)
+    df_bins = df_bins.with_columns([
+        pl.col("bin_start").cast(pl.Int64),
+        pl.col("bin_stop").cast(pl.Int64)
+    ])
+    df_bins_by_sample_hits = df_bins_by_sample_hits.with_columns([
+        pl.col("bin_start").cast(pl.Int64),
+        pl.col("bin_stop").cast(pl.Int64)
+    ])
+    df_bins_by_read_hits = df_bins_by_read_hits.with_columns([
+        pl.col("bin_start").cast(pl.Int64),
+        pl.col("bin_stop").cast(pl.Int64)
+    ])
+
```

Comment:
It would be good to retain the writing code in the CLI action

If any of these are anticipated to be large, then writing as gzip would be nice.

Reply:
...if any of these contain columns which are not a primitive type then writing as parquet would be better as it explicitly encodes type data. CSV is an untyped free-for-all, and not all parsers resolve the data the same particularly for non-primitives. 

If unfamiliar with that terminology, a primitive type would be e.g. int, float. Whereas an example of a non-primitive type is `List[int]`.

Reply:
differing this to a later PR because the current tsvs are easy to open and look at for testing and exploration purposes. It can get big though so I want to zip them and drop repeated info (e.g. list of samples) at one point.